text/3730-nit-lint-level.md

```diff
@@ -0,0 +1,402 @@
+
+These can also be implemented manually with any setup.
+For example, in a large, corporate, legacy C++ code base that [epage](https://github.com/epage/)
+worked with, a linter used a similar database to track what lints were "new" and reported those back to
+[ReviewBoard](https://www.reviewboard.org/).
+
+## Choosing lint levels
+
+When creating a lint or overriding a default lint level,
+its important to consider the semantics, something like:
+
+`error` is for hard errors that even prevent reporting of errors further in the build, running of tests, etc.
+This should be reserved for when there is no point building or testing anything further because of how broken the code is.
+
+`warn` generally has the semantics of a soft-error.
+Its non-blocking locally but will be blocked when the code is merged upstream.
+This should be used when eventual correctness or consistency is desired.
+
+`nit` is for coding suggestions for users that may or may not improve the code or may not need to be done yet.
+
+`allow` is for when there is no right answer (e.g. mutually exclusive lints), there are false positives, the lint is overly noisy, or the lint is of limited value.
+
+*Note:* This is for communicating the intent of this RFC and to get people started with this new feature and is not meant to precisely redefine the
+[lint levels](https://rustc-dev-guide.rust-lang.org/diagnostics.html#diagnostic-levels).
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+## Rustc
+
+Rustc focuses on the mechanics of a new, visible lint level `nit`.
+
+Add support for the new lint:
+```rust
+#![nit(deprecated)]
+```
+```console
+$ rustc --nit deprecated ...
+```
+```console
+$ rustc -Ndeprecated ...
+```
+
+These will be rendered like other lint levels with a diagnostic level and coloring that conveys the non-blocking, helpful nature.
+For example, this could be rendered like a `note:` or `help:` diagnostic level.
+
+Rustc will have a dynamic lint group of all `nit`s, much like `warnings` is all lints at level `warn` at the time of processing, `-Anits`.
+
+## Cargo
+
+Cargo implements the semantics for this to be a first-class non-blocking lint level.
+
+Add support for the new lint:
+```toml
+[workspace.lints.rust]
+deprecated = "nit"
+```
+
+A new config field will be added to mirror `build.warnings`:
+
+### `build.nits`
+
+- Type: string
+- Default: `allow`
+- Environment: `CARGO_BUILD_NITS`
+
+Controls how Cargo handles nits. Allowed values are:
+
+- `nit`: warnings are emitted as warnings.
+- `allow`: warnings are hidden (default).
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+Rust-analyzer might be limited in what it can do through LSP to avoid overwhelming the user with `nit`s.
+
+This requires a non-standard CI setup to report these messages.
```

Comment:
A follow-up RFC or compiler MCP could propose to make `--error-format=sarif` a first class citizen in `rustc`, replacing `clippy-sarif`, so that a CI setup with sarif error reporting becomes more the norm (or at least a standard option).

That was requested from Clippy before https://github.com/rust-lang/rust-clippy/issues/8121 but I had to push back on this, because properly supporting this could only be done in `rustc`.

Reply:
Discussed in 5e307601